File: library/sync-tasks/src/main/java/org/cru/godtools/sync/task/FollowupSyncTasks.java
package org.cru.godtools.sync.task;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.RestrictTo;

import org.ccci.gto.android.common.db.Query;
import org.ccci.gto.android.common.jsonapi.model.JsonApiObject;
import org.cru.godtools.model.Followup;
import org.cru.godtools.model.Language;

import java.io.IOException;
import java.util.List;

import retrofit2.Response;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public final class FollowupSyncTasks extends BaseSyncTasks {
    private static final Object LOCK_FOLLOWUPS = new Object();

    public FollowupSyncTasks(@NonNull final Context context) {
        super(context);
    }

    public void syncFollowups() throws IOException {
        synchronized (LOCK_FOLLOWUPS) {
            final List<Followup> followups = mDao.get(Query.select(Followup.class));
            for (final Followup followup : followups) {
                final Language language = mDao.find(Language.class, followup.getLanguageCode());
                followup.setLanguage(language);
                followup.stashId();

                final Response<JsonApiObject<Followup>> response = mApi.followups.subscribe(followup).execute();
                if (response.code() == 204) {
                    followup.restoreId();
                    mDao.delete(followup);
                }
            }
        }
    }
}

File: library/sync-tasks/src/main/java/org/cru/godtools/sync/task/FollowupSyncTasks.kt
package org.cru.godtools.sync.task

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.ccci.gto.android.common.db.Query
import org.ccci.gto.android.common.db.find
import org.cru.godtools.model.Followup
import org.greenrobot.eventbus.EventBus
import org.keynote.godtools.android.db.GodToolsDao
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FollowupSyncTasks @Inject internal constructor(private val dao: GodToolsDao, eventBus: EventBus) :
    BaseSyncTasks(eventBus) {
    private val followupMutex = Mutex()

    suspend fun syncFollowups() = withContext(Dispatchers.IO) {
        followupMutex.withLock {
            coroutineScope {
                dao.get(Query.select<Followup>())
                    .map { followup ->
                        async {
                            try {
                                followup.languageCode?.let { followup.setLanguage(dao.find(it)) }
                                followup.stashId()
                                api.followups.subscribe(followup).isSuccessful
                                    .also {
                                        if (it) {
                                            followup.restoreId()
                                            dao.delete(followup)
                                        }
                                    }
                            } catch (e: IOException) {
                                false
                            }
                        }
                    }.all { it.await() }
            }
        }
    }
}
